contracts/CourtPresaleActivate.sol

```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        require(ERC20(_token).safeTransferFrom(msg.sender, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+
+        // get the Uniswap exchange for the contribution token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(_token);
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        // swap tokens
+        address contributionTokenAddress = address(presale.contributionToken());
+        ERC20 token = ERC20(_token);
+        require(token.safeApprove(address(uniswapExchange), _amount), ERROR_TOKEN_APPROVAL_FAILED);
+        uint256 contributionTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, contributionTokenAddress);
+
+        // buy in presale
+        _buyAndActivate(msg.sender, contributionTokenAmount, contributionTokenAddress);
+
+        // refund leftovers if any
+        _refund(token);
+        _refund(contributionTokenAddress);
+    }
+
+    /**
+    * @dev It will send the received ETH to Uniswap to get presale contribution tokens,
+    *      convert the obtained tokens in the presale instance, and activate the converted tokens in the
+    *      jurors registry instance of the Aragon Court.
+    * @param _minTokens Minimum amount of presale contribution tokens obtained in Uniswap
+    * @param _deadline Transaction deadline for Uniswap
+    */
+    function contributeEth(uint256 _minTokens, uint256 _deadline) external payable {
+        require(msg.value > 0, ERROR_ZERO_AMOUNT);
+
+        address contributionTokenAddress = address(presale.contributionToken());
+
+        // get the Uniswap exchange for the contribution token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(contributionTokenAddress);
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        // swap tokens
+        uint256 contributionTokenAmount = uniswapExchange.ethToTokenSwapInput.value(msg.value)(_minTokens, _deadline);
+
+        // buy in presale
+        _buyAndActivate(msg.sender, contributionTokenAmount, contributionTokenAddress);
+
+        // make sure there's no ETH left
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            (bool result,) = msg.sender.call.value(ethBalance)("");
+            require(result, ERROR_ETH_REFUND);
+        }
+    }
+
+    function _buyAndActivate(address _from, uint256 _amount, address _token) internal {
         // approve to presale
-        require(ERC20(_token).safeApprove(address(presale), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+        require(ERC20(_token).safeApprove(address(presale), _amount), ERROR_TOKEN_APPROVAL_FAILED);
 
         // buy in presale
         presale.contribute(address(this), _amount);
```

Comment:
I don't have much context on this, but if we have the `bondedToken` in storage, is there a reason we're using `presale.contributionToTokens(_amount)` instead of `bondedToken.balanceOf(this)`?

Are they equivalent or are we protecting for the case where someone accidentally sends the bonded token to this contract (because the `presale` separately tracks the contribution amount from this contract)?

Reply:
Yes, they should be equivalent, unless someone accidentally sends as you said. Anyway, those would be sent on `_refund` later. Not sure what's the best option here unless we implement an escape hatch mechanism governed by AN DAO or something like that.